File: a-proofreading-tool.py
def is_palindrome(word):
    if not word:
        return False
    cleaned = word.lower()
    return cleaned == cleaned[::-1]


def find_palindrome_breaks(words):
    if not words:
        return []
    return [i for i, w in enumerate(words) if not is_palindrome(w)]


def find_repeated_phrases(words, phrase_length):
    if not words or phrase_length >= len(words) or phrase_length <= 0:
        return []

    phrases = []
    max_start = len(words) - phrase_length

    for i in range(max_start + 1):
        phrase_str = " ".join(words[i : i + phrase_length]).lower()
        phrases.append(phrase_str)

    counts = {}
    for p in phrases:
        counts[p] = counts.get(p, 0) + 1

    repeated_indices = []
    for i, p in enumerate(phrases):
        if counts[p] > 1:
            repeated_indices.append(i)

    return repeated_indices


def analyze_texts(texts, phrase_length):
    if not texts:
        return []

    result = []
    for words in texts:
        result.append(
            {
                "repeatedPhrases": find_repeated_phrases(words, phrase_length),
                "palindromeBreaks": find_palindrome_breaks(words),
            }
        )
    return result

#variant



def is_plindrome_two_pointers(string_val):
    if not string_val:
        return False
    cleaned = string_val.lower()
    left = 0
    right = len(cleaned) -1

    while left < right:
        if cleaned[left] != cleaned[right]:
            return False
        left += 1
        right -= 1

    return True

print(is_plindrome_two_pointers("racecar"))
print(is_plindrome_two_pointers("hello"))

File: a-proofreading-tool.js
function isPalindrome(word) {
  if (!word) return false;
  const cleaned = word.toLowerCase();
  return cleaned === cleaned.split("").reverse().join("");
}

function findPalindromeBreaks(words) {
  if (!words || words.length === 0) return [];
  const breaks = [];
  for (let i = 0; i < words.length; i++) {
    if (!isPalindrome(words[i])) {
      breaks.push(i);
    }
  }
  return breaks;
}

function findRepeatedPhrases(words, phraseLength) {
  if (!words || phraseLength >= words.length || phraseLength <= 0) return [];

  const phrases = []; 
  const maxStart = words.length - phraseLength;

  for (let i = 0; i <= maxStart; i++) {
    const phraseStr = words.slice(i, i + phraseLength).join(" ").toLowerCase();
    phrases.push(phraseStr);
  }

  const counts = {};
  for (const p of phrases) {
    counts[p] = (counts[p] || 0) + 1;
  }

  const repeatedIndices = [];
  for (let i = 0; i < phrases.length; i++) {
    if (counts[phrases[i]] > 1) {
      repeatedIndices.push(i);
    }
  }

  return repeatedIndices;
}

function analyzeTexts(texts, phraseLength) {
  if (!texts || texts.length === 0) return [];

  return texts.map(words => ({
    repeatedPhrases: findRepeatedPhrases(words, phraseLength),
    palindromeBreaks: findPalindromeBreaks(words)
  }));
}

//

function isPalindromeAdvanced(word) {
    if (!word) return false;

    const cleaned = word.toLowerCase().replace(/[^a-z0-9]/g, "");
    return cleaned === cleaned.split("").reverse().join("");
}
console.log(isPalindromeAdvanced("A man, a plan, a canal: Panama")); 

//

function isPalindromeTwoPointers(word) {
    if(!word) return false;
    const cleaned = word.toLowerCase();
    let left = 0;
    let right = cleaned.length -1;

    while (left < right) {
        if (cleaned[left] !== cleaned[right]) {
            return false;
        }
        left++;
        right--;
    }
    return true;
}

//

function isPalindromeJapanese(str) {
  let cleaned = str
    .replace(/[っ]/g, "つ")
    .replace(/[ゃ]/g, "や")
    .replace(/[ゅ]/g, "ゆ")
    .replace(/[ょ]/g, "よ");

  return cleaned === cleaned.split("").reverse().join("");
}

console.log(isPalindromeJapanese("しんぶんし")); // true
console.log(isPalindromeJapanese("きしぶしき")); 
